phabricatoremails/progress.py

```diff
@@ -0,0 +1,48 @@
+import time
+from dataclasses import dataclass
+from logging import Logger
+from typing import Callable
+
+from phabricatoremails.db import Db
+from phabricatoremails.model.persistence import QueryPosition
+from sqlalchemy.orm import Session
+
+
+@dataclass
+class PhabricatorProgress:
+    _logger: Logger
+    _poll_gap_seconds: int
+
+    def process(self, db: Db, pipeline: Callable[[Session, str], str]):
+        while True:
```

Comment:
The `Progress` abstraction is defined more in the [readme](https://github.com/mozilla-conduit/phabricator-emails/tree/1609266-phab-emails#progress). I agree that the term "progress" isn't ideal, I didn't have any better terms jump out at me. Let me know if you have any ideas!
(The two hardest problems in software are ... :wink:)

-----

> How does one exit this while loop? What happens when this process is killed or if an unexpected exception occurs?

You don't exit the while loop, it iterates continually, sending emails. The database transaction of each loop ensures that unexpected issues don't corrupt database state.
I'm adding this ^ to the class documentation.

Reply:
> You don't exit the while loop, it iterates continually, sending emails. The database transaction of each loop ensures that unexpected issues don't corrupt database state.

What happens when you want to redeploy your service? There should be a way to gracefully exit this loop without relying on the database transactions, etc...

> The Progress abstraction is defined more in the readme. I agree that the term "progress" isn't ideal, I didn't have any better terms jump out at me. Let me know if you have any ideas!

Though the readme can be very hand for someone figuring out how to use this application or develop it, it's not a substitute for a good docstring!

Reply:
> What happens when you want to redeploy your service? There should be a way to gracefully exit this loop without relying on the database transactions, etc...

When you want to redeploy your service, you stop the process, move your service, start the process.
Usually services are stopped with signals (e.g.: `SIGINT`/`SIGTERM`, then `SIGKILL` within seconds if it doesn't respond).
This gracefully handles such a signal: immediately rolls back the transaction, if any, then shuts down.

You could do some extra work like making sure you're finished rendering/sending emails when you receive a signal, but this sounds like overkill (and besides, when if you reach that limit and you get force killed?). How would you handle this differently?

Reply:
>You could do some extra work like making sure you're finished rendering/sending emails when you receive a signal, but this sounds like overkill (and besides, when if you reach that limit and you get force killed?). How would you handle this differently?

This is definitely not overkill and should be handled. `SIGINT`/`SIGTERM` should trigger your graceful shutdown. Finish the current iteration then exit. The database rollback should be a fail safe, and is not equivalent to a graceful shutdown.

Reply:
The tangible drawback of not handling `SIGINT`/`SIGTERM` is that we may accidentally double-send emails if the service stops during the mail-sending part of the pipeline.

The solution of waiting for our batch of emails to send won't resolve this issue, since rendering and sending the emails may be slow enough that we reach the `SIGKILL` timeout. Therefore, attempting the finish the current iteration isn't guaranteed to be more graceful than rolling back the current iteration and stopping.

Either way, the chances of the service being killed while sending emails (as opposed to sleeping) are relatively low. I think that there _are_ improvements we can make (make each "unit of work" smaller so that we're committing a valid state immediately after each individual emails sends), but I think that we should report the bug and prioritize in the future accordingly.

Reply:
> The tangible drawback of not handling `SIGINT`/`SIGTERM` is that we may accidentally double-send emails if the service stops during the mail-sending part of the pipeline.

It is also a better way of implementing a service and a general good practice.

> The solution of waiting for our batch of emails to send won't resolve this issue, since rendering and sending the emails may be slow enough that we reach the `SIGKILL` timeout. Therefore, attempting the finish the current iteration isn't guaranteed to be more graceful than rolling back the current iteration and stopping.

Has this been tested? Do you have more specific numbers here? If you're referring to a kubernetes grace period, you can configure this per service, and should be set to a reasonable timeout based on how long you expect a single iteration of the loop to finish. You should design the application so that it can be gracefully shut down when needed. Fail safes should be reserved for unexpected and unintentional issues. Also note that your exception handling will not be triggered if your application received certain kill signals (i.e. `session.rolback()`)

> Either way, the chances of the service being killed while sending emails (as opposed to sleeping) are relatively low. I think that there _are_ improvements we can make (make each "unit of work" smaller so that we're committing a valid state immediately after each individual emails sends), but I think that we should report the bug and prioritize in the future accordingly.

This is important to implement from the get go. We shouldn't rely on chance when there's going to be hundreds of users being affected by something predictable such as having to stop a service. Since this is a brand new service that is untested in a production setting, it would be essential to put as many safeguards in place to be able to handle things gracefully.

Reply:
> Has this been tested? Do you have more specific numbers here?

This is exactly why I _don't_ think waiting for the current iteration is a good idea - Amazon SES doesn't define a "maximum time" to send emails. We can set a reasonable grace period to `SIGKILL`, but we still won't be able to guarantee that it's long enough (what if SES is under heavy load and we get a _huge_ delay spike halfway through sending our emails?).

Trying to wait for your unit of work to finish before shutdown isn't a better practice than cleaning up and immediately shutting down _precisely because_ it adds uncertainty to your clean shutdown.

> If you're referring to a kubernetes grace period, ...

Exactly, I'm referring to the "grace period" concept that's similar between kubernetes/systemd/supervisord/etc.

> Also note that your exception handling will not be triggered if your application received certain kill signals (i.e. session.rolback())

That's the thing - when using a transaction, if you don't make it to `commit()`, it will be automatically rolled back. Many transaction-providing systems implement such that disconnecting will roll back any incomplete transactions. That's the beauty of transactions: they allow you to define a "critical section" of work that will either complete all together, or not at all.

> This is important to implement from the get go. We shouldn't rely on chance when there's going to be hundreds of users being affected by something predictable such as having to stop a service.

In this case, then we should be doubling down even harder on making this system resilient to shutdowns, whether they're scheduled or **unexpected/fatal.**

> It is also a better way of implementing a service and a general good practice.

This was the start of your comment, but I wanted to address it after all the context written above^. If we're focusing on system stability, then we need to optimally handle the "unexpected shutdown case". If we can handle the case when we stop the service at a moment's notice without losing data/double-sending emails, then we don't gain any benefit by finishing the "unit of work" on a scheduled shutdown (other than sending one more batch of emails right now, rather than on next startup, but that's negligible). Therefore, it's not clear to me how it's better practice to finish the unit of work.

-----

Finally, and most importantly: it's worth considering the impact of a failure here. On average, there's 1-2 events created per minute on Phabricator. If it takes a second (over-estimating) to handle these events (read from Phab, render the templates, inline CSS, encode as MIME, send the emails), then there's less than a 1/60 chance of stopping the application while it's handling events. Additionally, the sending the emails is a _fraction_ of handling the events, so that probability is even significantly smaller. Finally, assuming the shutdown is timed _perfectly_ so that we've sent some emails, but the loop doesn't finish, the impact is that a couple emails are going to be re-sent on-next startup. The chance of failure and the impact of failure are _incredibly low_ that it's probably safe to defer this work.

Reply:
> Handling a basic graceful shutdown is straightforward, and it should be part of this initial PR. A more advanced graceful shutdown can be added later as part of deferred work. At the very least, I would simply catch SIGTERM/SIGINT and change a variable indicating whether this service is running or not, rather than relying on while True. Something like Progress.is_running.

:+1:, we're on the same page about how to generally finish a unit of work before shutdown, I think the implementation details of this solution are clear.

> Based on SES documentation (https://docs.aws.amazon.com/ses/latest/APIReference/API_SendEmail.html) the API request is actually handled immediately and the email is queued. You should be getting an immediate response from send_email.

Based on that documentation, is it possible to define a grace period that will be long enough to finish sending emails _100% of the time_, regardless of network delays, AWS internal hiccups and unexpected circumstance?

Reply:
I don't think that this work should block this PR, but in the interest of time I'm going to implement the "finish unit of work" logic when handling a graceful shutdown.